bokeh/cli/core.py

```diff
@@ -0,0 +1,285 @@
+            instance. Syntax convention for this option is as follows:
+            <output_type>://<type_arg>
+
+            where:
+              - output_type: 'file' or 'server'
+              - 'file' type options: path_to_output_file
+              - 'server' type options syntax: docname[@url][@name]
+
+            Defaults to: --output file://cli_output.html
+
+            Examples:
+                --output file://cli_output.html
+                --output file:///home/someuser/bokeh_rocks/cli_output.html
+                --output server://clidemo
+
+            Default: file://cli_output.html.
+        title (str, optional): the title of your chart.
+            Default: None.
+        plot_type (str, optional): charts classes to use to consume and
+            render the input data.
+            Default: Line.
+        series (str, optional): Name of the series from the input source
+            to include in the plot. If not specified all source series
+            will be included.
+            Defaults to None.
+        palette (str, optional): name of the colors palette to use.
+            Default: None.
+        index (str, optional): Name of the data series to be used as index
+            when plotting. By default the first series found on the input
+            file is taken
+            Default: None
+        buffer (str, optional): if is `t` reads data source as string from
+            input buffer using StringIO(sys.stdin.read()) instead of
+            reading from a file or an url.
+            Default: "f"
+        sync_with_source (bool, optional): if True keep the charts source
+            created on bokeh-server sync'ed with the source acting like
+            `tail -f`.
+            Default: False
+
+
+        Attributes:
+            source (obj): datasource object for the created chart.
+            chart (obj): created chart object.
+        """
+        self.input = input_source
+        self.series = series
+        self.index = index
+        self.last_byte = -1
+        self.sync_with_source = sync_with_source
+
+        self.source = self.get_input(input_source, buffer)
+        # get the charts specified by the user
+        self.factories = create_chart_factories(plot_type)
+
+        if palette:
+            print ("Sorry, custom palettes not supported yet, coming soon!")
+
+        # define charts init parameters specified from cmd line and create chart
+        self.chart_args = get_chart_params(title, output)
+
+    def run(self):
+        """ Start the CLI logic creating the input source, data conversions,
+        chart instances to show and all other niceties provided by CLI
+        """
+        try:
+            self.chart = create_chart(
+                self.series, self.source, self.index, self.factories, **self.chart_args
+            )
+            self.chart.show()
+
+            self.has_ranged_x_axis = 'ranged_x_axis' in self.source.columns
```

Comment:
self.columns is needed when streaming data (and calling update_source). I also needed self.has_ranged_x_axis at some point but recently removed that. It probably could be a local variable but I would do it in a later PR. There are several things to be added and I'm not sure if we'll need it or not.